code/frameworks/angular/src/client/angular-beta/utils/BootstrapLock.ts

```diff
@@ -0,0 +1,50 @@
+/* eslint-disable no-await-in-loop */
+import { ApplicationRef } from '@angular/core';
+
+const queue: Array<() => Promise<void>> = [];
+let isProcessing = false;
+
+/**
+ * Reset compiled components because we often want to compile the same component with
+ * more than one NgModule.
+ */
+const resetCompiledComponents = async () => {
+  try {
+    // Clear global Angular component cache in order to be able to re-render the same component across multiple stories
+    //
+    // References:
+    // https://github.com/angular/angular-cli/blob/master/packages/angular_devkit/build_angular/src/webpack/plugins/hmr/hmr-accept.ts#L50
+    // https://github.com/angular/angular/blob/2ebe2bcb2fe19bf672316b05f15241fd7fd40803/packages/core/src/render3/jit/module.ts#L377-L384
+    const { ɵresetCompiledComponents } = await import('@angular/core');
+    ɵresetCompiledComponents();
+  } catch (e) {
+    /**
+     * noop catch
+     * This means angular removed or modified ɵresetCompiledComponents
+     */
+  }
+};
+
+export const queueBootstrapping = (fn: () => Promise<ApplicationRef>): Promise<ApplicationRef> => {
+  return new Promise<ApplicationRef>((resolve, reject) => {
+    queue.push(() => fn().then(resolve).catch(reject));
+
+    if (!isProcessing) {
+      processQueue();
+    }
+  });
+};
+
+const processQueue = async () => {
+  isProcessing = true;
+
+  while (queue.length > 0) {
+    const bootstrappingFn = queue.shift();
+    if (bootstrappingFn) {
+      await bootstrappingFn();
+      await resetCompiledComponents();
```

Comment:
> By moving this here, doesn't it prevent anyone from having the option of implementing a renderer that doesn't reset compiled components?

You're right. In `@storybookjs/testing-angular,` I haven't found where you are inheriting from `AbstractRenderer,` though. I guess you mentioned, in general, that people might use it to implement a custom renderer, which inherits from `AbstractRenderer`. I used the GitHub search, and I was not able to find any custom renderer inheriting `AbstractRenderer`. Could you find some to figure out whether this is a real problem? But I would guess we can avoid calling `resetCompiledComponents` in the lock function. Do you have an idea how we might avoid it?

Reply:
I am not sure how to best avoid it, at the moment. That can maybe wait for another PR, if necessary.

When I mentioned `@storybookjs/testing-angular`, I was just using it as an example, but forgot the renderer in that package isn't actually a render and doesn't extend `AbstractRenderer`. It is basically the renderer, without the bootstrapping part, because I create a component that can render the story and let the user's chosen test framework handle using it. So, this actually doesn't apply to it.